client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -168,24 +170,39 @@ private String updateGroupStateIfNeeded() throws ReaderNotInReaderGroupException
             groupState.checkpoint(atCheckpoint, getPosition());
             releaseSegmentsIfNeeded();
         }
-        atCheckpoint = groupState.getCheckpoint();
-        if (atCheckpoint != null) {
-            log.info("{} at checkpoint {}", this, atCheckpoint);
-            if (groupState.isCheckpointSilent(atCheckpoint)) {
+        String checkpoint = groupState.getCheckpoint();
+        if (checkpoint != null) {
+            log.info("{} at checkpoint {}", this, checkpoint);
+            if (groupState.isCheckpointSilent(checkpoint)) {
                 // Checkpoint the reader immediately with the current position. Checkpoint Event is not generated.
-                groupState.checkpoint(atCheckpoint, getPosition());
-                atCheckpoint = null;
+                groupState.checkpoint(checkpoint, getPosition());
+                if (atCheckpoint != null) {
+                    //In case the silent checkpoint held up releasing segments
+                    releaseSegmentsIfNeeded();
+                    atCheckpoint = null;
+                }
                 return null;
             } else {
+                atCheckpoint = checkpoint;
                 return atCheckpoint;
             }
+        } else {
+            atCheckpoint = null;
+            acquireSegmentsIfNeeded();
+            return null;
         }
-        acquireSegmentsIfNeeded();
-        return null;
     }
 
     @GuardedBy("readers")
     private void releaseSegmentsIfNeeded() throws ReaderNotInReaderGroupException {
+        for (Iterator<Segment> iterator = sealedSegments.iterator(); iterator.hasNext();) {
```

Comment:
This is enforcing this invariant, yes?

```
No portion of keyspace that is owned by one reader will
be read by another without the first reader checkpointing
first.
```

Reply:
There are two parts to that invariant:
The `handleEndOfSegment` method 2 lines down enforces that no checkpoints are pending. 
The fact that a checkpoint was just performed is enforced by this method's one and only caller.